### DISCUSSION TITLE
How to extend the type definition on `expect` in typescript

### DISCUSSION THREAD

**Discussion by iambinlin** [vitest-dev/vitest#1863] [Q&A] [answered]
For Example, In `setupFiles`: 

```ts
expect.extend({
    toBeNotNull(received, expected) {
        const { isNot } = this;
        return {
            pass: received !== null,
            message: () => `${received} is${isNot ? ' not' : ''} not null`,
        };
    },
});
```

then in `vite.config.ts`:

```ts
import { defineConfig } from 'vitest/config';
import AutoImport from 'unplugin-auto-import/vite';

export default defineConfig({
    test: {
        globals: true,
        transformMode: {
            web: [/\.[jt]sx$/],
        },
        setupFiles: './fixtures/vi.setup.ts',
    },
    plugins: [
        AutoImport({
            imports: ['vitest'],
            dts: '.autoImport/ai.d.ts', // generate TypeScript declaration
        }),
    ],
});

```

<img width="924" alt="image" src="https://user-images.githubusercontent.com/39019913/184916459-900cef57-987a-452b-ab01-02aba61f4491.png">

Since the type definition on `expect` is automatically generated by `unplugin-auto-import/vite`, how do I customize the extension of my type. Thanks a lot!

**Comment by pelletier197** [comment]
I have been struggling to find an answer to this. I tried many things but none of them works. Has anyone found a way?

**Comment by AriPerkkio** [comment, marked as answer]
I have one working example here: https://github.com/AriPerkkio/extend-to-be-announced/blob/98b1d51b3e8895a3872aff2668bbd67b96058bd1/src/vitest/register.ts#L15-L19. 
This is equivalent for Jest's extension: https://github.com/AriPerkkio/extend-to-be-announced/blob/98b1d51b3e8895a3872aff2668bbd67b96058bd1/src/jest/register.ts#L5-L18.

The documentation should be edited to include Typescript additions too. Currently there seems to be no mention of it here: https://vitest.dev/guide/extending-matchers.html.

**Reply by pelletier197** [reply]
Thanks for that. I managed to make it work this way too. That should definitely be part of the documentation you shared. Is anyone planning on opening a PR to add that? Otherwise I can do it.

**Reply by AriPerkkio** [reply]
Documentation PRs are always appreciated - feel free to open one @pelletier197 once you have time to do that. You can also ping me in the PR for a review.

**Comment by pelletier197** [comment]
> Documentation PRs are always appreciated - feel free to open one @pelletier197 once you have time to do that. You can also ping me in the PR for a review.

For sure, I will open it as soon as I can 🙂

**Comment by iambinlin** [comment]
<img width="929" alt="image" src="https://user-images.githubusercontent.com/39019913/215714438-84ec1bfd-7c6b-4b41-8d1d-8983ca54ad56.png">

<img width="892" alt="image" src="https://user-images.githubusercontent.com/39019913/215714518-5423e9bb-e30d-47f3-b22e-f41e566d7376.png">

This method can be solved. Thank you very much for your valuable  @AriPerkkio
